jwst/guider_cds/guider_cds.py

```diff
@@ -0,0 +1,133 @@
+    divided by the effective integration time.  If the mode is ID, the last 
+    group minus the first group is calculated for both integrations; the count 
+    rate is then given by the minimum of these two values for each pixel, 
+    divided by the effective integration time.  For the FINEGUIDE mode, the 
+    count rate is the average of the last 4 groups minus the average of the 
+    first 4 groups, divided by the effective integration time.  
+
+    Parameters
+    ----------
+    model: data model
+        input data model, assumed to be of type FGSModel
+    """
+    # get needed sizes and shapes
+    imshape, n_int, effinttm, readpatt = get_dataset_info(model)
+
+    if readpatt == 'ID': # force output to have single slice
+        new_model = datamodels.GuiderCdsModel((1,)+imshape)
+    else:
+        new_model = datamodels.GuiderCdsModel()
+
+    new_model.dq = model.pixeldq
+    slope_int_cube = np.zeros(( n_int,) + imshape, dtype = np.float64)
+
+    # loop over data integrations
+    for num_int in range(0, n_int):
+        data_sect = model.data[num_int, :,:,:]
+
+        if readpatt == 'FINEGUIDE':
+            first_4 = data_sect[:4,:,:].mean(axis=0)
+            last_4  = data_sect[-4:,:,:].mean(axis=0)
+            slope_int_cube[num_int,:,:] = last_4 - first_4
+
+        elif readpatt == 'ID':
+            grp_last = data_sect[1, :, :]
+            grp_first = data_sect[0, :, :]
+
+            if num_int == 0:
+                diff_int0 = grp_last - grp_first
+            if num_int == 1:
+                diff_int1 = grp_last - grp_first
+
+        else:  # ACQ1, ACQ2, or TRACK
+            grp_last = data_sect[1, :, :]
+            grp_first = data_sect[0, :, :]
+            slope_int_cube[ num_int,:,: ] = grp_last - grp_first
+        
+    if readpatt == 'ID':
+        new_model.data[0,:,:] = np.minimum( diff_int1, diff_int0 )/effinttm
+
+    else:  # FINEGUIDE, ACQ1, ACQ2, or TRACK
+        new_model.data = slope_int_cube/effinttm
+
+    new_model.update(model)  # ... and add all keys from input
+
+    return new_model
+
+
+def get_dataset_info(model):
+    """
+    Short Summary
+    -------------
+    Extract values for the number of reads, the number of pixels,
+    dataset shapes, the number of integrations, the instrument name,
+    the frame time, and the observation time??????????
+
+    Parameters
+    ----------
+    model: instance of Data Model
+       DM object for input
+
+    Returns
+    -------
+    imshape: (int, int) tuple
+       shape of 2D image
+
+    n_int: int
+       number of integrations
+
+    effinttim: float
+       effective integration time
+
+    readpatt: string
+        readpattern ...number of groups per integration
+
+
+    """ 
+    instrume = model.meta.instrument.name
+    frame_time = model.meta.exposure.frame_time
+    ngroups = model.meta.exposure.ngroups
+
+    effinttm = model.meta.exposure.integration_time
+    readpatt = model.meta.exposure.readpatt
```

Comment:
This could now be upgraded to use the EXP_TYPE keyword (model.meta.exposure.type), since we've convinced the SDP folks to include the EXP_TYPE keyword in guider products.